### ISSUE TITLE
Accessing the map from a computed property in the controller

### ISSUE THREAD

**Issue by radcliff** [miguelcobain/ember-leaflet#10]
Is it possible to access to the underlying `map` object directly? 

I'd like to be able to call a method like [`#getBoundsZoom()`](http://leafletjs.com/reference.html#map-getboundszoom) directly from the controller, in order to compute the `zoom` for the map. It doesn't work to put the call in an action triggered by `onLoad` because `zoom` will need to be resolved before the `onLoad` event can be fired.

**Comment by miguelcobain** [comment]
@radcliff what exactly are you trying to do?
If you're trying to set the bounds of your map, you can use the `bounds` property in the latest ember-leaflet version.
Interchangeable with `center`.

**Comment by radcliff** [comment]
Right. At first I thought I would like to use `bounds` all the time, but I need to fit the map to a collection of markers. So I am able to derive the bounds from the geojson collection of markers:

`bounds = L.geoJson(geojson).getBounds()`

but when the collection only contains *one* point, the bounds that is calculated makes no sense -- the same point is used for both corners. So setting the `bounds` with that will break the map. So in the case of the collection only having a single point, I need to pass a `center` and `zoom`, which is also fine. But now i have code duplication that is really ugly:

```
{{#if hasOneFeature}}
  {{#leaflet-map center=center zoom=zoom}}
    {{tile-layer url=tileLayer}}
    {{#each features as |feature|}}
      {{marker-layer location=feature.latlng icon=icon}}
    {{/each}}
  {{/leaflet-map}}

{{else}}
  {{#leaflet-map bounds=bbox}}
    {{tile-layer url=tileLayer}}
    {{#each features as |feature|}}
      {{marker-layer location=feature.latlng icon=icon}}
    {{/each}}
  {{/leaflet-map}}
{{/if}}
```
What I would like to be able to do is just stick with `center` and `zoom` in both cases, which will work if i can calculate the appropriate zoom for the bounding box via `getBoundsZoom()` -- but that is actually a method on the map (makes sense, needs to know the dimensions of the map container) and then you actually pass in the `bounds` as an argument:

`zoom = map.getBoundsZoom(bounds)  // might return 10 or something`

Getting the center is easy and not an issue:

`center = bounds.getCenter()`

**Comment by miguelcobain** [comment]
@radcliff exactly. I see your problem. What I'm saying is that you can use both center and bounds interchangeably. Example:

```
{{#leaflet-map center=center zoom=zoom bounds=bbox}}
  {{tile-layer url=tileLayer}}
  {{#each features as |feature|}}
    {{marker-layer location=feature.latlng icon=icon}}
  {{/each}}
{{/leaflet-map}}
```

When you set `center` the map centers on that point.
If after you set `bbox` the map sets the view that contains that bounding box.

You can't set both `center` and `bounds`, though (makes sense). But I think you can work that.

If you really want to call `getBoundsZoom`, I'd try to do that on the `onLoad` event using `e.target` which is an instance to the leaflet map. But please try the first approach.

**Comment by radcliff** [comment]
>You can't set both center and bounds, though (makes sense).

In the example you are showing both `center` and `bounds` params present at the same time, so I am confused as to whether you can use both params or not. It doesn't seem to work for me when I use both, is there something else to it?

**Comment by miguelcobain** [comment]
@radcliff You're right, I made it look a bit confusing.

What I meant was that you can't *create* a map with bot bounds and center.
But, at a later time, you can set bounds or center.

Since we're using DDAU (data down actions up), we communicate with our component by changing data.
If we want to know when the component changed the data, we use actions.
I just made an assertion that initial state must not have bounds and center set.

So, in short:
- make sure that initially only `center`(or `lat`/`lng`) or `bounds` is set
- if, at a later time, you want to set bounds, just `set` the bound property
- similar thing for `center`

Does this help?

**Comment by radcliff** [comment]
yes, thank you. so the way it works is that i can say:

```
{{#leaflet-map onViewreset=( action 'setBounds') bounds=bounds center=center zoom=zoom}}
  {{tile-layer url=tileLayer}}
  {{#each features as |feature|}}
    {{marker-layer location=feature.latlng icon=icon}}
  {{/each}}
{{/leaflet-map}}
```

`bounds` is not initially set in the controller, but this wont break the map initialization. THEN `onViewreset` I call my own `setBounds` function, that basically just checks if there is at least 2 features before setting the `bounds`. Works :hammer: :construction_worker: 

Here is what some of the controller looks like:

```
  bbox: Ember.computed 'geojson', () ->
    geojson = @get('geojson')
    return L.geoJson(geojson).getBounds()

  center: Ember.computed 'features', 'bbox', () ->
    bbox = @get('bbox')
    return bbox.getCenter()

  zoom: 15

  actions:
    setBounds: () ->
      features = @get('features')  # from geojson.features

      if features.length > 1
        bbox = @get('bbox')
        @set('bounds', bbox)
```

@miguelcobain Thanks again for the help!

**Comment by miguelcobain** [comment]
Perhaps using the `onLoad` action would be more idiomatic, wouldn't it?
